### PR TITLE
feat: implement sitemap generation

### PR DESCRIPTION
**Pull request by harrysolovay**
Utilizes Capi to do so

### CODE REVIEW

File: capi/src/write/sitemap.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  for (const [, page] of ctx.pageBySrcPath.entries()) {
+    routes.set(page.route, true);
+    if (page.versions) {
+      Object.entries(page.versions).forEach(([, version]) =>
+        routes.set(version, true),
+      );
+    }
+  }
+
+  const src = `<?xml version="1.0" encoding="utf-8" standalone="yes" ?>
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:29]: What does standalone="yes" do?

File: capi/src/write/sitemap.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  return ` <url>
+    <loc>http://docs.amplify.aws${route}</loc>
+    <lastmod>${yyyyMmDd}</lastmod>
+    <changefreq>hourly</changefreq>
+    <priority>.5</priority>
+  </url>`;
+}
+
+export async function sitemap(config: t.Config, ctx: t.Ctx): Promise<void> {
+  const routes = new Map<string, true>();
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:18]: Why a map here instead of an array?

Comment (harrysolovay) [reply capi/src/write/sitemap.ts:18]: It's fast, and ensures no duplicates

File: capi/src/write/sitemap.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+  }
+
+  const src = `<?xml version="1.0" encoding="utf-8"?>
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+  ${[...routes.keys()].map(createSitemapRoute).join("\n")}
+</urlset>
+`;
+
+  const destPathPieces = config.outDir.split(path.sep);
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:35]: If something in the route assignment broke it seems we would just generate an empty sitemap file. Should we throw some kind of error instead?

Comment (harrysolovay) [reply capi/src/write/sitemap.ts:35]: Maybe I don't understand correctly...

destPathPieces being of length 0 won't happen with the Amplify Console environment... and if anyone is running Capi locally, and they have the project at their filesystem root (before library, before user)... then they're begging for trouble.

File: capi/src/write/sitemap.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    if (page.versions) {
+      Object.entries(page.versions).forEach(([, version]) =>
+        routes.set(version, true),
+      );
+    }
+  }
+
+  if (routes.size > 0) {
+    const src = `<?xml version="1.0" encoding="utf-8"?>
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:31]: Can we fix the formatting here

File: capi/src/write/sitemap.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  ${[...routes.keys()].map(createSitemapRoute).join("\n")}
+</urlset>
+`;
+
+    const destPathPieces = config.outDir.split(path.sep);
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:36]: If `outDir` is undefined this will break

File: capi/src/write/sitemap.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    destPathPieces.pop();
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:37]: what is this `pop` doing?

File: capi/src/write/sitemap.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    routes.set(page.route, true);
+    if (page.versions) {
+      Object.entries(page.versions).forEach(([, version]) =>
+        routes.set(version, true),
+      );
+    }
+  }
+
+  if (routes.size > 0) {
+    const src = `<?xml version="1.0" encoding="utf-8"?>
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:30]: Let's use uppercase UTF-8

File: capi/src/write/sitemap.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import * as t from "../types";
+import * as path from "path";
+import * as fs from "fs-extra";
+
+const now = new Date();
+const yyyyMmDd = `${now.getFullYear()}-${now.getMonth()}-${now.getDate()}`;
+
+function createSitemapRoute(route: string): string {
+  return ` <url>
+    <loc>http://docs.amplify.aws${route}</loc>
```
Comment (jordanranz) [inline capi/src/write/sitemap.ts:10]: This should be `https`

File: capi/src/write/sitemap.ts (hunk 5 above)
Comment (harrysolovay) [reply capi/src/write/sitemap.ts:36]: `outDir` cannot be undefined

File: capi/src/write/sitemap.ts (hunk 3 above)
Comment (harrysolovay) [reply capi/src/write/sitemap.ts]: Moved this into an optional util in the config

File: capi/src/write/sitemap.ts (hunk 4 above)
Comment (harrysolovay) [reply capi/src/write/sitemap.ts:32]: Fixed

File: capi/src/write/sitemap.ts (hunk 6 above)
Comment (harrysolovay) [reply capi/src/write/sitemap.ts]: Moved this into config

File: capi/src/write/sitemap.ts (hunk 8 above)
Comment (harrysolovay) [reply capi/src/write/sitemap.ts]: Fixed

File: capi/src/write/sitemap.ts (hunk 7 above)
Comment (harrysolovay) [reply capi/src/write/sitemap.ts]: Done

File: tasks/build.ts [hunk 9]
```diff
@@ -43,4 +43,9 @@ c.API({
   outDir: "client/src/api",
   publicDir: "client/src/assets",
   watch,
+  getSitemapDestDir(config) {
```
Comment (jordanranz) [inline tasks/build.ts:46]: why not just `appDir: "client/src"`

Comment (harrysolovay) [reply tasks/build.ts:46]: My thinking is it should be based on the configured dest path.

**Pull request merged**